Add file reader interface and XYZ reader implementation, take 3

berquist opened the pull request:
Continuing work from https://github.com/cclib/cclib/pull/454 and https://github.com/cclib/cclib/pull/502.

langner commented on cclib/io/__init__.py:
```diff
@@ -7,12 +7,14 @@
 
 """Contains all writers for standard chemical representations"""
 
+
```
nit: maybe just one line

langner commented on cclib/io/filereader.py:
```diff
@@ -0,0 +1,44 @@
+
+
+class Reader(object):
+    """Abstract class for reader objects."""
```
Do we really want it to be abstract? If so, let's use `abc`.

berquist replied:
I think that should be in a separate PR, since it needs to be done for the writers as well, along with tests for each.

langner commented on cclib/io/filereader.py:
```diff
@@ -0,0 +1,44 @@
+        Inputs:
+          source - A single filename, stream [TODO], or list of filenames/streams [TODO].
+        """
+
```
nit: don't need the blank line here

langner commented on cclib/io/xyzreader.py:
```diff
@@ -0,0 +1,83 @@
+        while True:
+
+            try:
+
```
nit: unneeded extra blank line

langner commented on test/io/testxyzreader.py:
```diff
@@ -0,0 +1,116 @@
+        self.assertEqual(data.atomnos.shape, atomnos.shape)
+        self.assertTrue(np.all(np.equal(data.atomnos, atomnos)))
+
+        atomcoords = np.array([[
```
Can the test be formulated in a way that doesn't require replicating the data?

berquist replied:
Not if you want to test for explicit correctness, no.

langner replied:
Well, what if we wrote it back out and made sure the resulting file was the same?

berquist replied:
I started to write a roundtrip test, and I still think it's important to have, but I don't know that it solves this problem. The reason is because we need to fix the precision in the output file, and there are many reasons why something like the MD5 of the input and output files would differ. In that case, you could compare the generated output to a reference generated output.

Is that what you're referring to?

langner replied:
The numbers in that example don't look weird to get mangled due to rounding or whatever. I wouldn't compare MD5's myself, just the exact strings, but I think it can be done in many ways. I do think not having the data in the actual test would be better, though. If you think it's not worth the effort at this time, that's OK with me as well.

langner commented on test/io/testxyzreader.py:
```diff
@@ -0,0 +1,116 @@
+        xyz = cclib.io.xyzreader.XYZ(fpath)
+        data = xyz.parse()
+
+        attrs = ('natom', 'atomnos', 'atomcoords')
```
You can say the same thing on 3 lines explicitly. Also, since you're actually testing these attribute below, I don't think you need to assert their existence.

langner commented on test/io/testxyzreader.py:
```diff
@@ -0,0 +1,116 @@
+
+        self.assertEqual(data.atomcoords.shape, atomcoords.shape)
+        self.assertTrue(np.all(np.equal(data.atomcoords, atomcoords)))
+
```
nit: maybe 2 blank lines before invocation of main?

langner commented on test/io/testxyzreader.py:
```diff
@@ -0,0 +1,109 @@
+        ])
+
+        self.assertEqual(data.atomcoords.shape, atomcoords.shape)
+        self.assertTrue(np.all(np.equal(data.atomcoords, atomcoords)))
```
I don't think you need to test the shape if you're checking equivalence. Also, perhaps we should use something from `numpy.testing` here?

langner commented on test/io/testxyzreader.py:
```diff
@@ -0,0 +1,109 @@
+        """Is an XYZ file with a two geometries read into a ccData properly?
+        """
+        fpath = os.path.join(__filedir__, "data/uracil_two.xyz")
+        assert os.path.exists(fpath)
```
I don't think we need to assert additionally here. If the file doesn't exist, the test already fails.

berquist commented:
I added the round-trip tests against reference XYZ outputs and removed the hard-coded coordinates, but left in the atomic numbers in `testxyzreader`. Let me know if I should remove those also.

I think that only having the round-tripping tests is closer to integration testing than unit testing, so we should also have something simpler.

langner commented:
> I added the round-trip tests against reference XYZ outputs and removed the hard-coded coordinates, but left in the atomic numbers in testxyzreader. Let me know if I should remove those also.
> 
> I think that only having the round-tripping tests is closer to integration testing than unit testing, so we should also have something simpler

Great, I think this is a good compromise. This way, even if the the specific coordinates get changed in the file, the test won't need to be updated.
